src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1159,12 +1162,47 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     nameof(expression));
             }
 
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            var tryIncludingPlaceholder = false;
+            switch (inputType)
```

Comment:
Actually, the only guaranteed thing about the `inputType` is that it controls where `value` is found. Instead of all of this, do the following after the existing `switch`:
``` c#
    string typeString = tagBuilder.Attributes["type"];
    if (_placeholderInputTypes.Contains(suppliedTypeString))
    {
        modelExplorer = modelExplorer ??
            ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
        if (modelExplorer.Metadata.Placeholder != null)
        {
            tagBuilder.MergeAttribute("placeholder", modelExplorer.Metadata.Placeholder);
        }
    }
```